crates/types/src/traits/signature_key.rs

```diff
@@ -139,3 +139,53 @@ pub trait SignatureKey:
     #[must_use]
     fn genesis_proposer_pk() -> Self;
 }
+
+pub trait BuilderSignatureKey:
+    Send
+    + Sync
+    + Clone
+    + Sized
+    + Debug
+    + Hash
+    + Serialize
+    + for<'a> Deserialize<'a>
+    + PartialEq
+    + Eq
+    + PartialOrd
+    + Ord
+    + Display
+{
+    /// The type of the keys builder would use to sign its messages
+    type BuilderPrivateKey: Send
+        + Sync
+        + Sized
+        + Clone
+        + Debug
+        + Eq
+        + Serialize
+        + for<'a> Deserialize<'a>
+        + Hash;
+
+    /// The type of the signature builder would use to sign its messages
+    type BuilderSignature: Send
+        + Sync
+        + Sized
+        + Clone
+        + Debug
+        + Eq
+        + Serialize
+        + for<'a> Deserialize<'a>
+        + Hash;
+
+    /// Type of error that can occur when signing data
+    type SignError: std::error::Error + Send + Sync;
+
+    /// validate the message with the builder's public key
+    fn validate_builder_signature(&self, signature: &Self::BuilderSignature, data: &[u8]) -> bool;
+
+    /// sign the message with the builder's private key
+    fn sign_builder_message(
+        private_key: &Self::BuilderPrivateKey,
+        data: &[u8],
+    ) -> Result<Self::BuilderSignature, Self::SignError>;
```

Comment:
I don't know enough to review the full PR (though it seems fine to me), but since I was thinking about this in other places recently: could this take `self` rather than `private_key` as an argument? 

As it is currently, you wouldn't be able to use `sign_builder_message`  without having direct access to the private key, which forces any struct that implements `BuilderSignatureKey` to expose the private key. But for security, the private key should probably be a private field.

I think this would also mean `BuilderPrivateKey` doesn't need to be an associated type.

(Unfortunately HotShot already propagates private keys unnecessarily as it is, so if it's not an easy thing to change here this is probably fine relative to that)

Reply:
what about having &self, and have a different `fn get_privatekey(&self)->&Self::BuilderPrivateKey`,
```rust
fn sign_builder_message(&self, data: &[u8])-> Result<....>{
private_key = self.get_private_key();
...
..
}
```

Reply:
IMO if you have an interface to produce signatures you already have full control of the private key for all practical purposes, whether it's some wrapper struct or not

Reply:
My concern is about inadvertently misusing/leaking it rather than maliciously signing with it. For example, `BuilderPrivateKey` has a `Serialize` bound, which means it's possible to write into an array and send it out.

It's not likely we'll ever make this specific mistake, but making the field private lets you rule out a lot of misuses at the source, and it's much easier to audit a limited trait interface than it is to make sure every use of an exposed private key in the entire codebase is correct.